### DISCUSSION TITLE
Obtain tree of halfspaces from nef polyhedron

### DISCUSSION THREAD

**Discussion by aothms** [CGAL/cgal#7632] [Q&A]
It is my understanding that a nef polyhedron is essentially conceptually a tree of half spaces. In the documentation on Nef polyhedra we read for example

> ... we can evaluate a CSG-tree with halfspaces ...

https://doc.cgal.org/latest/Nef_3/index.html

Given the implementation in CGAL. How can I obtain this tree of halfspaces from an Nef_polyhedron?

- I look at visit_shell_objects() but I didn't seem to uncover the actual hierarchy using this approach
- I've also seen SMexplorer() but I haven't been able to understand how it is to be used.

Any help is greatly appreciated.

**Comment by sloriot** [comment]
As soon as the nef polyhedron is built, the only information we have are sphere maps at each vertex of the Nef (each vertex holds an arrangement on the sphere of all incidents simplices). So, AFAIK, it is not possible to get the tree.

**Reply by aothms** [reply]
Many thanks for the quick reply.

I think I'm going to try to reconstruct such a tree then from the sphere maps:

- Loop over the shalfedges to get facet to facet connectivity. These should be in ccw order so judging from the plane normal should be quick to assess whether concave or convex.

- Accumulate such pairs for the entire polyhedron and find connected components of convexly connected facets, which becomes a branch in a tree.

- Proceed somehow (have to work it out still) until a tree is built

The reason for this is the following:

- Take in a large amount of non-precise polyhedra

- "Snap" / align half-spaces so that facets across (different) polyhedra align / are co-planar

- Rebuild new nefs using set complement+difference on the snapped half-spaces

- Perform boolean union on these new nefs, resulting in a cleaner result compared to operating on the original unaligned nefs: less small facets, less small gaps between almost touching facets in the imprecise inputs.

--

If none of this makes sense, this would be a good time to tell me :)

**Reply by sloriot** [reply]
I understand the end goal, i.e. snapping parts that should be identical but that are not, so that the output mesh does not feature tiny volumes. What I don't understand is why you have to go with a Nef to extract the tree. From what I understand you already know the sequence of operations done using the input polyhedra. Maybe if you have a small example it would help understanding.
